test/integration/swaps.ts

```diff
@@ -0,0 +1,148 @@
+    const router = await getUniswapRouter(hre);
+
+    const amountIn = to_d18(0.0001);
+
+    const pathsPrices = await generatePaths(amountIn, weth.address, bdx.address);
+    console.log(
+      pathsPrices.map(x => ({
+        path: x.path,
+        amountOut: d18_ToNumber(x.amountOut)
+      }))
+    );
+
+    const bestPath = await chooseBestPath(pathsPrices);
+    console.log(d18_ToNumber(bestPath.amountOut));
+    expect(bestPath.path.length).to.eq(2);
+
+    weth.approve(router.address, amountIn);
+    const currentBlock = await hre.ethers.provider.getBlock("latest");
+    await router.swapExactTokensForTokens(amountIn, 0, bestPath.path, deployer.address, currentBlock.timestamp + 1e5);
+  });
+
+  it("Should choose best path (lenght 3)", async () => {
+    const deployer = await getDeployer(hre);
+
+    const bdx = await getBdx(hre);
+    const weth = await getWeth(hre);
+    const router = await getUniswapRouter(hre);
+
+    const currentBlock = await hre.ethers.provider.getBlock("latest");
+
+    // sabotage the price on the direct swap, so the algorithm prefers an indirect path
+    const amountInForSabotage = to_d18(10);
+    console.log("amountInForSabotage:", d18_ToNumber(amountInForSabotage));
+    await weth.approve(router.address, amountInForSabotage);
+    await router.swapExactTokensForTokens(amountInForSabotage, 0, [weth.address, bdx.address], deployer.address, currentBlock.timestamp + 1e5);
+
+    const amountIn = to_d18(0.0001);
+
+    const pathsPrices = await generatePaths(amountIn, weth.address, bdx.address);
+    console.log(
+      pathsPrices.map(x => ({
+        path: x.path,
+        amountOut: d18_ToNumber(x.amountOut)
+      }))
+    );
+
+    const bestPath = await chooseBestPath(pathsPrices);
+    console.log(d18_ToNumber(bestPath.amountOut));
+    expect(bestPath.path.length).to.eq(3);
+
+    weth.approve(router.address, amountIn);
+    await router.swapExactTokensForTokens(amountIn, 0, bestPath.path, deployer.address, currentBlock.timestamp + 1e5);
+  });
+
+  async function getTokensAvailebleForSwap() {
+    return [
+      ...(await getAllBDStables(hre)).map(x => x.address),
+      (await getBdx(hre)).address,
+      (await getWbtc(hre)).address,
+      (await getWeth(hre)).address
+    ];
+  }
+
+  async function generatePaths(amountIn: BigNumber, addressIn: string, addressOut: string): Promise<{ path: string[]; amountOut: BigNumber }[]> {
+    const availableTokens = await getTokensAvailebleForSwap();
+    const midTokens = availableTokens.filter(x => x !== addressIn && x != addressOut);
+    const paths = [[addressIn, addressOut], ...midTokens.map(x => [addressIn, x, addressOut])];
+
+    const router = await getUniswapRouter(hre);
+    const pathsPrices = [];
+
+    for (const path of paths) {
+      let amountsOut;
+      try {
+        amountsOut = await router.getAmountsOut(amountIn, path);
+      } catch (_) {
+        continue; // handle unsupported paths like [wrbtc -> eths -> bdx]
```

Comment:
I'm not sure if this is the right approach... It's generic, but:

1. We make pointless rpc calls (it's slow, we're wasting rpc calls)
2. What if somebody creates wrbtc-eths pool in our system (random pools don't bother us - only whitelisted tokens take part in this selection)? Can such pool be harmful to the user?

Maybe we should just hardcode possible paths instead?

Reply:
I think we should improve the logic behind paths generation to allow only whitelisted pairs. this will help us avoid RBTC-ETHs-BDX since RBTC-ETHs is not a whitelisted pair. 
it will also remove invalid paths like RBTC-RBTC-BDX

Reply:
thanks for the suggestion, now it's way better

Reply:
Awesome PR @AndrzejGisWideLab ! This makes things much clearer.
This code I believe is already whitelisted, isn't it?
You're getting the links from `getAvailableSwapLinks` which gets it from `getPools`.
`getPools` uses `getAllBDStables` which only takes the stables from our own whitelisted list of stables.
There is no generic call for the router to get the pools here, so I believe we're good (unless I missed something).

Maybe worth just mentioning it in a comment.

Reply:
On a general business note, we'll have to decide as part of https://lagoslabs.atlassian.net/browse/LAGO-554 which smart routes we would like to support. But let's discuss this one offline :)

Reply:
> This code I believe is already whitelisted, isn't it?
@kazazor Yes, the latest version builds routes based on whitelisted links